### DISCUSSION TITLE
Needed to add "upload_port = /dev/ttyUSB0" to platformio.ini - now SOURCE CONTROL in Code complains

### DISCUSSION THREAD

**Discussion by daveclifford** [DCC-EX/EX-T3-WiFi#79] [General]
I had to add an upload_port directive (/dev/ttyUSB0) to the platformio.ini file to get the upload to work in the  EX-T3-Wifi git directory and now Source Control in Code wants me to commit my change.

Do you know where I can add the upload_port directive in another place outside the git tree that will provide the directive globally?

Dave

**Comment by alex-code** [comment]
That's odd, thought it tried to find the port automatically.

Maybe environment variables?
https://docs.platformio.org/en/latest/envvars.html
https://docs.platformio.org/en/latest/envvars.html#uploading

**Comment by daveclifford** [comment]
Alex,

Thanks for showing me the way............................

I am using Linux and it seems the upload port in the Lnux version is not
being automagically found.

Linux uses shells (my distribution uses the bash shell) to set up the
session environment for a user when initially logged in.

Using your first reference [
https://docs.platformio.org/en/latest/envvars.html],

# Unix (bash, zsh)

export VARIABLE_NAME=VALUE

shows the export command to set a variable to be used during the session.

Platformio uses a prefix to identify its variables, so the following
statement should work:

*export PLATFORMIO_UPLOAD_PORT=/dev/ttyUSB0*

I will remove the reference I put in the platformio.ini file
[*upload_port = /dev/ttyUSB0]  *and set it in the session file.

I will report back later on my success and what worked; I hope it helps others.

Dave

**Reply by alex-code** [reply]
If it works I can look at adding a tips/hints bit to the docs for Linux users.

**Reply by daveclifford** [reply]
Put the following  line  in my session .bashrc 

**export PLATFORMIO_UPLOAD_PORT=/dev/ttyUSB0**

and it worked fine!!!

Benefit of this addition in the session .bashrc is that it does NOT change the platformio.ini file in the EX-T3-WiFi repository and cause a highlight of  the difference between my working copy and the github version.


**"If it works I can look at adding a tips/hints bit to the docs for Linux users."**

Anything you can do to help the users (MAC and Linux) will be greatly appreciated.!!!!  We want this process to be as plug-and-play as possible.

----------------------------------------------------------------------------

[Redacted  several superfluous lines of log  shown by several lines of dots]

Looking for upload port...
**Using manually specified: /dev/ttyUSB0**
Uploading .pio/build/firebeetle2/firmware.bin
esptool.py v4.4
Serial port /dev/ttyUSB0
Connecting.....
Chip is ESP32-D0WD-V3 (revision v3.0)
.
.
.
Changing baud rate to 460800
Changed.
.
.
Configuring flash size...
Flash will be erased from 0x00001000 to 0x00005fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x0000e000 to 0x0000ffff...
Flash will be erased from 0x00010000 to 0x00155fff...
Compressed 17408 bytes to 12112...
Writing at 0x00001000... (100 %)
Wrote 17408 bytes (12112 compressed) at 0x00001000 in 0.6 seconds (effective 219.6 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 134...
Writing at 0x00008000... (100 %)
Wrote 3072 bytes (134 compressed) at 0x00008000 in 0.1 seconds (effective 300.6 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Writing at 0x0000e000... (100 %)
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.2 seconds (effective 419.9 kbit/s)...
Hash of data verified.
Compressed 1333120 bytes to 797747...
Writing at 0x00010000... (2 %)
Wrote 1333120 bytes (797747 compressed) at 0x00010000 in 19.5 seconds (effective 547.7 kbit/s)...
Hash of data verified.
Leaving...
Hard resetting via RTS pin.

**Comment by daveclifford** [comment]
"If it works I can look at adding a tips/hints bit to the docs for Linux users"

I'm in the process of writing exaclty that - How do I format the file for you - such as an .md file?

Lead the way

Dave

**Reply by alex-code** [reply]
Hi Dave,

This page has good docs on how to write docs 😁
https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

If you do it in VSCode you can have a side by side editor/preview.
https://code.visualstudio.com/docs/languages/markdown#_markdown-preview
